helix-core/src/main/java/org/apache/helix/controller/stages/PerReplicaThrottleStage.java

```diff
@@ -220,6 +271,274 @@ private MessageOutput throttlePerReplicaMessages(IdealState idealState,
+          expectedStateCountByPartition.get(partition);
+      Map<String, Integer> partitionCurrentStateCounts =
+          currentStateCountsByPartition.get(partition);
+
+      List<Message> partitionMessages = selectedResourceMessages.get(partition);
+      if (partitionMessages == null) {
+        continue;
+      }
+
+      // sort partitionMessages based on transition priority and then creation timestamp for transition message
+      // TODO: sort messages in same partition in next PR
+      Set<String> disabledInstances =
+          cache.getDisabledInstancesForPartition(resourceName, partition.getPartitionName());
+      for (Message msg : partitionMessages) {
+        if (!Message.MessageType.STATE_TRANSITION.name().equals(msg.getMsgType())) {
+          logger.debug(
+              "Event info: {} Message: {} not subject to throttle in resource: {} with type {}",
+              _eventId, msg, resourceName, msg.getMsgType());
+          continue;
+        }
+
+        boolean isUpward = !isDownwardTransition(stateModelDef, msg);
+
+        // TODO: add disabled disabled instance special treatment
+
+        String toState = msg.getToState();
+
+        // TODO: dropped and Error state special treatment
+
+        Integer minimumRequiredCount = partitionExpectedStateCounts.getOrDefault(toState, 0);
+        Integer currentCount = partitionCurrentStateCounts.getOrDefault(toState, 0);
+
+        //
+        if (isUpward && (currentCount < minimumRequiredCount)) {
+          recoveryMessages.add(msg);
+          // It is critical to increase toState value by one here. For example, current state
+          // of 3 replica in a partition is (M, O, O). Two messages here bringing up the two O to S.
+          // In this case, the first O->S would be classified as recovery. Then this line would
+          // increase the S state in partitionCurrentStateCounts value by one. Next O->S message
+          // would be correctly marked as load message.
+          partitionCurrentStateCounts.put(toState, currentCount + 1);
+        } else {
+          loadMessages.add(msg);
+        }
+      }
+    }
+  }
+
+  protected void applyThrottling(String resourceName,
+      StateTransitionThrottleController throttleController, StateModelDefinition stateModelDef,
+      boolean onlyDownwardLoadBalance, List<Message> messages, Set<Message> throttledMessages,
+      StateTransitionThrottleConfig.RebalanceType rebalanceType) {
+    boolean isRecovery =
+        rebalanceType == StateTransitionThrottleConfig.RebalanceType.RECOVERY_BALANCE;
+    if (isRecovery && onlyDownwardLoadBalance) {
+      logger.error("onlyDownwardLoadBalance can't be used together with recovery_rebalance");
+      return;
+    }
+
+    // TODO: add message sorting in next PR
+    logger.trace("throttleControllerstate->{} before load", throttleController);
+    for (Message msg : messages) {
+      if (onlyDownwardLoadBalance) {
+        if (!isDownwardTransition(stateModelDef, msg)) {
```

Comment:
nit, if (onlyDownwardLoadBalance && !isDownwardTransition(stateModelDef, msg))... for simplicity.

Reply:
changed.